src/main/kotlin/no/skatteetaten/aurora/boober/service/resourceprovisioning/S3StorageGridProvisioner.kt

```diff
@@ -0,0 +1,189 @@
+import no.skatteetaten.aurora.boober.service.openshift.OpenShiftResponse
+import no.skatteetaten.aurora.boober.service.openshift.OperationType
+import no.skatteetaten.aurora.boober.utils.convert
+import no.skatteetaten.aurora.boober.utils.normalizeLabels
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
+import org.springframework.stereotype.Service
+
+data class SgProvisioningRequest(
+    val tenant: String,
+    val objectAreaName: String,
+    val deploymentName: String,
+    val targetNamespace: String,
+    val bucketPostfix: String
+)
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+data class StorageGridCredentials(
+    val tenantName: String,
+    val serviceEndpoint: String,
+    val accessKey: String,
+    val secretKey: String,
+    val bucketName: String,
+    val objectPrefix: String,
+    val username: String,
+    val password: String,
+    val bucketRegion: String? = null
+)
+
+data class SgRequestsWithCredentials(
+    val request: SgProvisioningRequest,
+    val credentials: StorageGridCredentials
+)
+
+class S3ProvisioningException(message: String, cause: Throwable? = null) : ProvisioningException(message, cause)
+
+@Service
+@ConditionalOnProperty(value = ["integrations.s3.variant"], havingValue = "storagegrid", matchIfMissing = false)
+class S3StorageGridProvisioner(
+    val openShiftDeployer: OpenShiftDeployer,
+    val openShiftClient: OpenShiftClient,
+    val openShiftCommandService: OpenShiftCommandService,
+    val herkimerService: HerkimerService,
+    val operationScopeFeature: OperationScopeFeature,
+    @Value("\${integrations.storagegrid.bucket.region:us-east-1}") val defaultBucketRegion: String,
+    @Value("\${integrations.storagegrid.provisioning-timeout:20000}") val provisioningTimeout: Long,
+    @Value("\${integrations.storagegrid.provisioning-status-check-interval:1000}") val statusCheckIntervalMillis: Long
+) {
+    private val logger = KotlinLogging.logger { }
+
+    fun getOrProvisionCredentials(applicationDeploymentId: String, requests: List<SgProvisioningRequest>)
+            : List<SgRequestsWithCredentials> {
+
+        if (requests.isEmpty()) return emptyList()
+
+        provisionMissingObjectAreas(applicationDeploymentId, requests)
+        return herkimerService.getObjectAreaAdminCredentials(applicationDeploymentId, requests)
+    }
+
+    fun provisionMissingObjectAreas(applicationDeploymentId: String, requests: List<SgProvisioningRequest>) {
+        val existingResources = herkimerService.getObjectAreaResourcesIndex(applicationDeploymentId)
+        if (existingResources.isNotEmpty()) {
+            logger.debug("ObjectArea(s) ${existingResources.keys.joinToString()} was already provisioned for applicationDeploymentId=$applicationDeploymentId")
+        }
+        val missingObjectAreas = requests
+            .filter { !existingResources.containsKey(it.objectAreaName) }
```

Comment:
Kunne vi hatt en metode for denne filteret. Enten en extension eller vanlig metode. Jeg må lese to ganger og tenke litt for å se hva som skjer inni filteren her.